### ISSUE TITLE
Creating a Deterministic wallet using Mnemonics

### ISSUE THREAD

**Issue by Chibuzo** [bitpay/bitcore-wallet-client#368]
I am trying to create a deterministic wallet. I use Bip39 to generate the mnemonics for the users like this:

`var bip39 = require('bip39');
var mnemonic = bip39.generateMnemonic();
seed = bip39.mnemonicToSeed(mnemonic, 'user-password');`

My challenge is how to use the mnemonic or seed from the mnemonic to create a new wallet using the createWallet() method
`API.createWallet(walletName, copayerName, m, n, opts, opts.network, opts.walletPrivKey, opts.id, opts.withMnemonics, cb)`

There's little documentation for the parameters passed into that method.

**Comment by dabura667** [comment]
https://github.com/bitpay/bitcore-wallet-client/issues/296

This issue will help you a little.

tl;dr

1. instanciate the BWC object with your BWS URL etc. (Like the example in README)
2. load credentials from the mnemonic (or rather, since it looks like you are wanting to use the BIP39 passphrase feature, you might want to load from xprv.) `client.seedFromExtendedPrivateKey(xprvString)` should be what you want.
3. After loading the credentials, create a wallet using:

```javascript
client.createWallet("My Cool Wallet Name", "me", 1, 1, {}, function(err, ret) { console.log(ret) })
```
walletName is the name that appears for the wallet in your copay etc. copayerName is your name, irrelevant if you are making a single user wallet, default is "me" but you have to give a value. m=how many sigs needed, n=how many copayers (m=2;n=3; means 2 of 3 people need to sign to send. m=1;n=1; is normal single user wallet.) all the opts are not needed for my example.

**Comment by dabura667** [comment]
If the xprv has been created as a wallet on the BWS before, createWallet will give you an error... so instead use openWallet to fill in the blanks for your object.

**Comment by Chibuzo** [comment]
Thank you @dabura667 for your response.
I stumbled on bitcore-mnemonics and decided to use it. Below is what I did:

`var mnemonic = new Mnemonic();`
`client.seedFromMnemonic(mnemonic, {network: 'testnet', passphrase: 'user-passphrase'});`
`client.createWallet("My Cool Wallet Name", "me", 1, 1, {network: 'testnet'}, function(err, ret) { console.log(ret) });`

It didn't work. I realized it doesn't execute `client.seedFromMnemonic()` successfully because when I put console.log('something') just below that line, it doesn't display.

I also tried using private key like this
`var private_key = mnemonic.toHDPrivateKey('user-passphrase');`
`client.seedFromExtendedPrivateKey(private_key)`;
It still didn't execute beyond that line.

I noticed that the key created when I use `mnemonic.toHDPrivateKey('user-passphrase')` always starts with 'xprv9...', but when I create wallet without trying to seed from mnemonics or private key, the private key returned after successful creation is usually shorter and doesn't start with 'xprv9'.

Any assistance would be appreciated.

**Comment by dabura667** [comment]
bwc only takes strings as arguments, not Mnemonic objects and HDPrivateKey objects.

iirc

mnObject.phrase will give you a string for mnemonic and .toString() should give you a string for HDprivatekey

**Comment by Chibuzo** [comment]
Hi @dabura667 , is it possible for a user change their passphrase maybe when it has been compromised?

**Comment by isocolsky** [comment]
Hi @Chibuzo, changing the passphrase will result in a different extended private key. This means you won't have access to your previous funds. I don't think this is what you are expecting?

**Comment by dabura667** [comment]
the "passphrase" should be thought of more as "the 13th word" in your phrase.

1. You can use your 12 words with no passphrase, and get wallet A.
2. You can use your 12 words with "apple" as the passphrase, and get wallet B.
3. You can use your 12 words with "orange" as the passphrase, and get wallet C.

There is no way to *change* the passphrase and get the same wallet.

**Comment by dabura667** [comment]
So if your 13th word is leaked..... just make sure they don't also have your 12 other words, and you will be fine.

**Comment by Chibuzo** [comment]
Thank you @isocolsky and @dabura667  for the explanation. I now understand very well.

So here is my challenge:
I'm using bitcore-wallet-client to build a bitcoin platform where users can have a bitcoin wallet. I felt it would be more efficient and secure to use mnemonics to open the wallet and carry out transactions since the users won't be running the service on their personal computer, so I don't save wallet credentials as user.dat files, I rather save their mnemonics on the database user table so when they successfully login using email and password, I'll pick their mnemonic and open their wallet. 
I know this isn't secure so I intend to encrypt each mnemonic using each user password as encryption key, but I'm still not sure if that would be a secure approach.

Please how can I secure users mnemonics so it can be used to open their wallet whenever they login to the app using their login credentials (email and password)?

**Comment by dabura667** [comment]
You can do anything you want. But you need to understand the risk for each decision.

1. If your server is hacked, the hacker has a list of all the mnemonics, and he doesn't need to know who's is who's, he can just check all of them at once for a balance.
2. If you encrypt the user's mnemonic using the user's password, then if they forget the password, you can not "reset the password" for them. (This is a common support request in Copay, many users don't write the phrase and forget their password)

It depends which you want to avoid... you can try avoiding both by making some compromise, but that's difficult to say.

**Comment by Chibuzo** [comment]
Thank you @dabura667 for the response. I wish you told me the compromise I could make so I would also deliberate on it, but I don't know how difficult it is to say so I won't bother you on that.

I would go for the second option. I would advice users to save their mnemonics properly in case they ever need to reset their password.

So please can you recommend a strong encryption algorithm I could use to encrypt the mnemonics?

Thank you.

**Comment by dabura667** [comment]
Use bcrypt for password hashing (it automatically salts too.)

Then hash the bcrypt result with SHA256 and use that data as the key for AES256.

**Comment by Chibuzo** [comment]
Thank you @dabura667 

I'll do just that.
